ARCHITECTURE.md

```diff
@@ -0,0 +1,133 @@
+
+### Dependency injection
+
+An important architectural component of this library is dependency injection, implemented here using [TSyringe](https://github.com/Microsoft/tsyringe). Dependencies are declared in `packages/*/src/dependencies.ts`.
+
+#### Declaring order
+
+Something important to realize is that the order in which the dependencies are declared (for a given container) matters. Let's have a look at some code to make things clearer.
+
+```
+container.register<IOidcHandler>("browser:oidcHandler", {
+  useClass: AggregateOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: RefreshTokenOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: AuthorizationCodeWithPkceOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: ClientCredentialsOidcHandler,
+});
+```
+
+Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler` and `ClientCredentialsOidcHandler` are its underlying handlers. When receiving a request, `AggregateOidcHandler` will first call to its instance of `RefreshTokenOidcHandler` to check if it can handle it. If so, the instance of `RefreshTokenOidcHandler` will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler` or `ClientCredentialsOidcHandler` will not be called. This means that it is important to declare the dependencies from the most specialized to the most generic, because if a fallback handler that can handle all requests is declared first, the other more specialized handlers will not be called.
+
+#### Mocks and tests
+
+Dependency injection makes the codebase more flexible, because it is only at runtime that each component will be presented with the dependencies it declared, which means it is easier to add a dependency to a component without changing the whole codebase.
+
+However, mocking dependency injection in test code wouldn't bring any value. Instead, the object we want to test can be constructed with mocked dependencies provided to its constructor. For instance, a class such as
+
+```
+@injectable()
+export default class RefreshTokenOidcHandler implements IOidcHandler {
+  constructor(
+    @inject("node:tokenRefresher") private tokenRefresher: ITokenRefresher,
+    @inject("node:storageUtility") private storageUtility: IStorageUtility
+  ) {}
+  // ...
+```
+
+can be tested as follows: 
+```
+const refreshTokenOidcHandler = new RefreshTokenOidcHandler(
+  someMockedTokenRefresher,
+  someMockedStorageUtility
+);
+```
+
+## Mapping OIDC flows to the code
+
```

Comment:
This is very useful - what might make it even more so is to add in which use cases which flow is used?

Reply:
That's certainly a possibility, however I'd like to have a discussion first: wouldn't that get rather deep into explaining OIDC/OAuth, which is information available in greater details elsewhere, and stray away from presenting how the codebase relates to OIDC/OAuth ?

Reply:
Hmm, I was thinking of a single comment relating it to Solid (e.g. "this is used by Solid apps running in the browser to login" and "this is used by Solid apps that run on the server or natively to maintain a session"), but perhaps I'm misunderstanding how specific to certain use cases these flows are and I should just read up on OIDC again :P

Reply:
Well it's more that I'm afraid I would give a truncated view of which flows may be used in which cases, while really I don't expect things to be Solid-specific here, but to hold in any OIDC context. However, if you think it may be useful, I could details which flows are explicitly NOT supported, e.g. the refresh flow in the browser, or explain why using the client credential flow is not something that works well with Solid.

Reply:
Actually I'll do that, because figuring out what things are absent is always trickier than finding the things that are there

Reply:
Yeah absolutely, sounds like a good idea.

Reply:
78e9ddd6ab6e555a1f559a7807465b4dd7e12cb1